File: Blind-75/SlidingWindow/MinWindowSubstring/pb-76.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(s) < len(t): return "" 
        window, countT = {}, {}
        for c in t:
            countT[c] = countT.get(c, 0) + 1
        need, have = len(countT), 0
        res, minLength = [-1, -1], float('inf')
        l = 0
        for r in range(len(s)):
            c = s[r]
            window[c] = window.get(c, 0) + 1
            if c in countT and window[c] == countT[c]:
                have += 1
            while have == need:
                if (r-l+1) < minLength:
                    res = [l,r]
                    minLength = r-l+1
                window[s[l]] -= 1
                if s[l] in countT and window[s[l]] < countT[s[l]]:
                    have -= 1
                l += 1
        l,r = res
        return s[l:r+1] if (minLength != float('inf')) else ""

File: Blind-75/SlidingWindow/MinWindowSubstring/pb-76.java
import java.util.*;
class Solution {
    public String minWindow(String s, String t) {
        if (s.length() < t.length()) return "";
        HashMap<Character, Integer> mapT = new HashMap<>();
        HashMap<Character, Integer> window = new HashMap<>();
        String res = ""; int lengthRes = Integer.MAX_VALUE;
        for (int i = 0; i < t.length(); i++){
            mapT.put(t.charAt(i), mapT.getOrDefault(t.charAt(i), 0) + 1);
        }
        int need = mapT.size(); int have = 0;
        int l = 0;
        for (int r = 0; r < s.length(); r++){
            char c = s.charAt(r);
            if (mapT.containsKey(c)){
                window.put(c, window.getOrDefault(c, 0)+1);
                if (window.get(c).equals(mapT.get(c))) have++;
            }    
            while (have == need){
                if ((r-l+1) < lengthRes){
                    lengthRes = r-l+1;
                    res = s.substring(l,r+1);
                }
                char leftChar = s.charAt(l);
                if (mapT.containsKey(leftChar)) {
                    window.put(leftChar, window.get(leftChar) - 1);
                    if (window.get(leftChar) < mapT.get(leftChar)) {
                        have--;
                    }
                }
                l++;
            }
        }
        return res;
    }
}